File: OptimalTransportNetworkToolbox_v.1.0.4b/Code/find_node.m
%{
==============================================================
 OPTIMAL TRANSPORT NETWORKS IN SPATIAL EQUILIBRIUM
 by P. Fajgelbaum, E. Schaal, D. Henricot, C. Mantovani 2017-19
 ================================================ version 1.0.4

[id]=find_node(graph,x,y): returns the index of the node closest to the 
coordinates (x,y) on the graph

Arguments:
- graph: structure that contains the underlying graph (created by
create_map, create_rectangle or create_triangle functions)
- x: x coordinate on the graph between 1 and w
- y: y coordinate on the graph between 1 and h

-----------------------------------------------------------------------------------
REFERENCE: "Optimal Transport Networks in Spatial Equilibrium" (2019) by Pablo D.
Fajgelbaum and Edouard Schaal.

Copyright (c) 2017-2019, Pablo D. Fajgelbaum, Edouard Schaal
pfajgelbaum@ucla.edu, eschaal@crei.cat

This code is distributed under BSD-3 License. See LICENSE.txt for more information.
-----------------------------------------------------------------------------------
%}

function id=find_node(graph,x,y)

distance=(graph.x-x).^2 + (graph.y-y).^2;
[~,id]=min(distance);

end % End of function

File: misc/plain matlab to julia (not working)/main/find_node.jl

"""
    find_node(graph, x, y)

Returns the index of the node closest to the coordinates (x,y) on the graph.

# Arguments
- `graph`: structure that contains the underlying graph (created by
create_map, create_rectangle or create_triangle functions)
- `x`: x coordinate on the graph between 1 and w
- `y`: y coordinate on the graph between 1 and h
"""
function find_node(graph, x, y)
    distance = (graph.x .- x).^2 + (graph.y .- y).^2
    _, id = findmin(distance)
    return id
end


# In this translation, I assumed that `graph` is a structure (or in Julia, a composite type) with fields `x` and `y` that are arrays of the same size. The `.-` operator is used for element-wise subtraction. The `findmin` function is used to find the minimum value in an array and its index, which is equivalent to the `min` function in Matlab. The underscore `_` is used to ignore the minimum value, as we are only interested in the index.